packages/expo-calendar/android/src/main/java/expo/modules/calendar/CalendarModule.kt

```diff
@@ -0,0 +1,988 @@
+package expo.modules.calendar
+
+import android.Manifest
+import android.content.ContentUris
+import android.content.ContentValues
+import android.content.Context
+import android.content.Intent
+import android.database.Cursor
+import android.os.Bundle
+import android.provider.CalendarContract
+import android.text.TextUtils
+import android.util.Log
+import expo.modules.interfaces.permissions.Permissions
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+import org.unimodules.core.ExportedModule
+import org.unimodules.core.ModuleRegistry
+import org.unimodules.core.ModuleRegistryDelegate
+import org.unimodules.core.Promise
+import org.unimodules.core.arguments.ReadableArguments
+import org.unimodules.core.errors.InvalidArgumentException
+import org.unimodules.core.interfaces.ExpoMethod
+import org.unimodules.core.interfaces.RegistryLifecycleListener
+import java.text.ParseException
+import java.text.SimpleDateFormat
+import java.util.*
+import kotlin.collections.ArrayList
+
+class CalendarModule(
+  private val mContext: Context,
+  private val moduleRegistryDelegate: ModuleRegistryDelegate = ModuleRegistryDelegate(),
+) : ExportedModule(mContext), RegistryLifecycleListener {
+  private val mPermissions: Permissions by moduleRegistry()
+
+  private val moduleCoroutineScope = CoroutineScope(Dispatchers.Default)
+  private val contentResolver
+    get() = mContext.contentResolver
+
+  private val sdf = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").apply {
+    timeZone = TimeZone.getTimeZone("GMT")
+  }
+
+  override fun getName(): String = "ExpoCalendar"
+
+  private inline fun <reified T> moduleRegistry() = moduleRegistryDelegate.getFromModuleRegistry<T>()
+
+  override fun onCreate(moduleRegistry: ModuleRegistry) {
+    moduleRegistryDelegate.onCreate(moduleRegistry)
+  }
+
+  override fun onDestroy() {
+    try {
+      moduleCoroutineScope.cancel(ModuleDestroyedException())
+    } catch (e: IllegalStateException) {
+      Log.e(TAG, "The scope does not have a job in it")
+    }
+  }
+
+  //region Exported methods
+  @ExpoMethod
+  fun getCalendarsAsync(type: String?, promise: Promise) {
+    if (!checkPermissions(promise)) {
+      return
+    }
```

Comment:
```suggestion
  private fun withPermissions(promise: Promise, block: ()->Unit) {
    if (!checkPermissions(promise)) {
      return
    }
    
  	block()  
  }
  
  @ExpoMethod
  fun getCalendarsAsync(type: String?, promise: Promise) = withPermissions(promise) {
```
You can think it we can inline this helper function 😉